### ISSUE TITLE
API support for partial orders among initializers

### ISSUE THREAD

**Issue by caesarxuchao** [kubernetes/kubernetes#50714]
### Current status
`metadata.initializers.pending` is a list of initializer names. Our [doc](https://kubernetes.io/docs/admin/extensible-admission-controllers/) says the initializer controller is triggered if its name matches `metadata.initializers.pending[0]`, that means initializers run in serial. The order is not enforced.

### Initializers need to run in parallel

Otherwise initialization is too slow. The watch latency can be a few seconds, an update request can take a few hundreds ms, also an initializer controller might take time to reserve external resources. Running initialization in serial won't scale to more than 10 initializers.

Initializer controllers should use PATCH to avoid conflicts.

### Initializers have dependencies

Not all initializers can run in parallel. Taking our existing mutating admsision plugins as an example, initializers that add/remove containers must finish before `alwayspullimages`, `resourcequota`, and `initialresource` initializers (assume they are converted to initializers).

### API requirements

* Cluster admin needs to be able to specify partial order for initializers. 
* The order must be dynamic. Admin can expand the order when adding new initializers.
* The apiserver should be able to enforce the order.

### Suggested API changes
For the [Initializer registration](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go#L62) API, add a `prerequisites` field:
```go
type Initializer struct {
	Name string
	// Prerequisites are the initializers that must finish before this initializer is triggered.
	// We can add an admission plugin to detect cycles, or let the intiailizer admission
	// plugin detect the circle (less efficient).
	Prerequisites []string
	Rules []Rule
	...
}
```

For the [initializer list](https://github.com/kubernetes/kubernetes/blob/v1.8.0-alpha.2/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L255) in each object, i.e., `metadata.initializers.pending[x]`, add the `prerequisites` field:

```go
type Initializer struct {
	Name string
	// Prerequisites are the initializers that must finish before this initializer.
	// When an object is created, the initializer admission plugin copies the prerequisites 
	// from the initializer registration API to here.
	// An update request is invalid if it removes this initializer while there is
	// still any prerequisite initializer in metadata.initializer.pending list.
	Prerequisites []string
}
```

@smarterclayton @erictune @deads2k

**Comment by erictune** [comment]
Can we just have a fixed set of stages, and everything is parallel within the stage, and we have advice about what behaviors each initializer can do in a given stage.
The stages I am thinking of are:

1. Adding: can add new repeated fields.  Example: anything that injects containers or volumes into containers.
2. Editing: can set unset fields to objects but not add repeated fields.  Example: things that set defaults on each container (needs all containers present)
3. Validating: may reject an object, but should not modify it.   Example: resourceQuota.

**Comment by erictune** [comment]
If we don't give some guidance about how to order things, it will be hard for users, and various initializers will be incompatible.

**Comment by caesarxuchao** [comment]
Benchmark set up:
* on a gke 1.7.3 alpha cluster
* The cluster has 1 node, with 8 vCPUs
* 10 initializers running in serial. 
* The initializer controllers were effectively no-op, they just patched the pod to remove the initializer from the pending list. 
* The initializer controllers were deployed as 10 containers in a single pod.
* The cluster didn't have other workload. 

Result:
* **The total creation and intialization time was ~200ms**.
* The apiserver log showed each patch only took <10 ms. 
* The arrival interval of the PATCH request was also <10ms, which meant that the watch latency was ~10ms.

The latency seems to be extremely low, given that our [SLO](http://blog.kubernetes.io/2016/03/1000-nodes-and-beyond-updates-to-Kubernetes-performance-and-scalability-in-12.html) for API responsiveness is "99% of all API calls return in less than 1s". @wojtek-t, as our performance expert, could you help estimate how bad these numbers would be in a busy cluster? How can I mimic a busy cluster? I guess creating extra pods don't help, I need to test a cluster with much more nodes.

**Comment by liggitt** [comment]
> given that our SLO for API responsiveness is "99% of all API calls return in less than 1s"

I don't disagree with exploring running initializers in parallel, but that SLO is being replaced by a better defined metric (see https://github.com/kubernetes/community/pull/881, which explicitly mentions initializer latencies)

**Comment by wojtek-t** [comment]
> A create call that blocks on initialization logically cannot have the same performance target as a single one of those initialization requests.

Yes - exactly. That was our goal. If a user defines a custom initializer/webhook or sth we have completely no control over it, so we can't give any guarantees about it.
That said we decided to proceed with the approach that we guarantee the SLO only in a default setup. that if a users adds his own initializer, we can basically substract the time of it from the total latency.

However, if we are going to use initializers by default (which I guess we may want a couple of them), they should be fast (because they are part of default setup then). But I think there won't be tens of them in a default setup.

>  The watch latency can be a few seconds

@caesarxuchao - is it based on some of your observations? It definitely shouldn't be the case. It should be in the order of tens of milliseconds (in heavily loaded clusters maybe small hundreds).

**Comment by caesarxuchao** [comment]
Thanks @liggitt @wojtek-t 

> The watch latency can be a few seconds

That's my observation when debugging flaky e2e tests. The number is for the worst cases (otherwise the e2e test will pass).

RE. SLO, @erictune suggested the goal to be reducing the lower bound of initializers latency, e.g., assuming initializers are no-ops, how fast can 10 initializers finish. In this perspective, we might need to enable parallel initializers.

If watch latency is small hundreds of ms, parallel initializers doesn't seem to be a priority for beta, as we expect there to be <10 initializers.

**Comment by erictune** [comment]
@liggitt clearly if the initializers are slow, we don't want to count that against the SLO.

On the other hand, if we end up shipping default initializers with Kubernetes, those need to fall under the SLO. I've heard people mention moving some things that are currently compiled in admission controllers to initializers.

**Comment by wojtek-t** [comment]
Correct - this is exactly what I wanted to express above saying:

> However, if we are going to use initializers by default (which I guess we may want a couple of them), they should be fast (because they are part of default setup then). But I think there won't be tens of them in a default setup.

**Comment by erictune** [comment]
Ah, sorry, Wojtek, yes, agree with what you already said.

**Comment by erictune** [comment]
@caesarxuchao do we have an estimate of how many initializers we expect to be default, a few versions from now?

**Comment by caesarxuchao** [comment]
I think we will convert an admission plugin to initializer only if it needs to be extensible. E.g., in the PersistentVolumeLabel case (#44680), we want each cloud provider to provide their own version of the intializer controller.

I only know we are going to convert podpreset to use initializer. I don't see use cases for other plugins. @smarterclayton @liggitt might have different ideas. 

Here is a list of mutating admission plugins in today's kubernetes:
* Limitranger (updates pod.spec.containers.resources)
* ResourceQuota
* PodNodeSelector (updates pod.node.selector)
* ServiceAccount (updates pod.spec.serviceaccountname pod.spec.volumes.secret)
* PodPreset
* DefaultTolerationSeconds (only changes pod.spec.tolerations)
* PodTolerationRestriction (only changes pod.spec.tolerations)
* AlwaysPullImages (only changes pod.spec.ImagePullPolicy)
* InitialResources (only changes pod annotation)
* PersistentVolumeLabel (already done, only cares about pvc)
* DefaultStorageClass (only cares about pvc)
* NamespaceAutoProvision (creates namespace)

**Comment by caesarxuchao** [comment]
Chatted with @smarterclayton on slack. We don't plan to move existing ones. We will add a few new ones,  "the ones for PVC, probably one for service IP allocation, pod preset". They care about different resources, so will never run in serial with each other.

**Comment by smarterclayton** [comment]
I do expect a number of plugins in the ecosystem to increase, but I suspect
there will be a ramp up curve.  4 max per resource in the next 6 months?

On Aug 17, 2017, at 9:43 PM, Chao Xu <notifications@github.com> wrote:

Chatted with @smarterclayton <https://github.com/smarterclayton> on slack.
We don't plan to move existing ones. We will add a few new ones, "the ones
for PVC, probably one for service IP allocation, pod preset". They care
about different resources, so will never run in serial with each other.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/50714#issuecomment-323237656>,
or mute the thread
